Source/FieldSolver/ElectrostaticSolver.cpp

```diff
@@ -346,17 +370,124 @@ WarpX::computePhiCartesian (const amrex::Vector<std::unique_ptr<amrex::MultiFab>
 
     // Solve the Poisson equation
     linop.setDomainBC( lobc, hibc );
+
+    for (int lev=0; lev < rho.size(); lev++){
+        rho[lev]->mult(-1._rt/PhysConst::ep0);
+    }
+
     MLMG mlmg(linop);
     mlmg.setVerbose(2);
     mlmg.setMaxIter(max_iters);
     mlmg.solve( GetVecOfPtrs(phi), GetVecOfConstPtrs(rho), required_precision, 0.0);
+}
+#endif
 
-    // Normalize by the correct physical constant
-    for (int lev=0; lev < rho.size(); lev++){
-        phi[lev]->mult(-1._rt/PhysConst::ep0);
+/* \bried Set Dirichlet boundary conditions for the electrostatic solver.
+
+    The given potential's values are fixed on the boundaries of the given
+    dimension according to the desired values from the simulation input file,
+    boundary.potential_lo and boundary.potential_hi.
+
+   \param[inout] phi The electrostatic potential
+   \param[in] idim The dimension for which the Dirichlet boundary condition is set
+*/
+void
+WarpX::setPhiBC( amrex::Vector<std::unique_ptr<amrex::MultiFab> >& phi,
+                 std::array<bool,AMREX_SPACEDIM> dirichlet_flag,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_lo,
+                 Array<amrex::Real,AMREX_SPACEDIM> phi_bc_values_hi ) const
+{
+    // check if any dimension has Dirichlet boundary conditions
+    bool has_Dirichlet = false;
+    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+        if (dirichlet_flag[idim]) {
+            has_Dirichlet = true;
+        }
     }
-}
+    if (!has_Dirichlet) return;
+
+    // loop over all mesh refinement levels and set the boundary values
+    for (int lev=0; lev <= max_level; lev++) {
+
+#ifdef AMREX_USE_OMP
+#pragma omp parallel if (amrex::Gpu::notInLaunchRegion())
 #endif
+        for ( MFIter mfi(*phi[lev], TilingIfNotGPU()); mfi.isValid(); ++mfi ) {
+            // Extract the potential
+            auto phi_arr   = phi[lev]->array(mfi);
+            // Extract tileboxes for which to loop
+            const Box& tb  = mfi.tilebox( phi[lev]->ixType().toIntVect() );
+
+            const auto lo = lbound(phi_arr);
+            const auto hi = ubound(phi_arr);
+
+            amrex::ParallelFor( tb,
+                [=] AMREX_GPU_DEVICE (int i, int j, int k) {
+                    int idx, idx_lo, idx_hi;
+
+                    for (int idim=0; idim<AMREX_SPACEDIM; idim++){
+                        // check if the boundary in this dimension should be set
+                        if (!dirichlet_flag[idim]) continue;
+
+                        if (idim == 0){
+                            idx = i;
+                            idx_lo = lo.x + 1;
```

Comment:
Having this single ParalleFor loop is good.

Why is this using `lo.x + 1`? Isn't the boundary at `lo.x`? Also, `lo` and `hi` are the bounds of the current tilebox which is not necessarily the bounds of the total domain. I would have expected something like this

amrex::Box const& domain = Geom(0).Domain();
const auto lo = lbound(domain);
const auto hi = ubound(domain);

Reply:
Thanks for pointing out the total domain versus current tilebox bounds differences, that clarifies things for me somewhat. I guess using the tilebox boundaries worked anyway since the cells that were not on the total domain boundary were overwritten during the fieldsolve. I've changed it though to only set cells on the total domain boundary. I could use some help to further understand the code behaviour here. When I queried the bounds (for the tilebox) before with
```
            const auto lo = lbound(phi_arr);
            const auto hi = ubound(phi_arr);
```
it would give `lo = (-1,-1,0)` and `hi = (65,9,0)` for a 64 x 8 cell domain when running on 1 processor, which is why I used `lo.x + 1` and `hi.x - 1`. Now that I switched to the total domain bounds using the code you provided above, I get `lo = (0,0,0)` and `hi = (63,7,0)`. So to get the same result as before I now use `lo.x` and `hi.x + 1`. This does seem to give the expected result, as shown in the attached plot which comes from the input file in `Examples/Tests/ElectrostaticDirichletBC`. If I set the hi potential on the cells with index `hi.x` (instead of `hi.x+1`), the potential is overwritten. Is this asymmetry in the handling of the lower and upper boundary correct?
![boundary_potentials](https://user-images.githubusercontent.com/40245517/113817290-f8970700-972a-11eb-8ac7-3f395008b88c.png)

Reply:
With a 64x8 sized test case, it almost certainly is using only one box so in that case the tilebox is the same as the whole domain. There is the parameter `amr.max_grid_size` that determines the size of blocks. Try setting this to 16 for example to divide the domain up into multiple blocks. At the beginning of the output from the run, it specifies the number of blocks used in the simulation.

Regarding lo and hi, the numbers are generally relative to the number of cells, hence the hi being 63. I would suggest using `domain_nodes = domain.surroundingNodes()` and take the lbound and ubound of domain_nodes.

Reply:
Another note - when using the whole domain, the index of the node at the lower boundary will always be zero.

Reply:
I have incorporated the `domain.surroundingNodes()` call as you suggested, which cleared up the question about having to use `hi.x + 1` to now just needing `hi.x`. I tested the implementation with setting `amr.max_grid_size = 16` (it seems to work fine) and changed the test for the dynamic boundary condition to have at least 2 boxes.